Function and IndexedBase inherits assumptions from Symbol

bsamseth opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
This PR is a continuation of PR #16085 resulting from a discussion with @smichr [there](https://github.com/sympy/sympy/pull/16085#discussion_r287332934).
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

If `Function` or `IndexedBase` is instantiated with a `Symbol` they will inherit the associated assumptions:

```python
from sympy import Symbol, Function, IndexedBase, log, exp, ask, Q
from sympy.abc import x

X = IndexedBase(Symbol("X", real=True))
f = Function(Symbol("f", real=True))

assert ask(Q.real(X))
assert log(exp(X)) == X
assert ask(Q.real(f(x)))
assert log(exp(f(x))) == f(x)
```  


#### Other comments

For `Function` this is currently only applied to `UndefinedFunction`, i.e. the result of `Function('f')`. This is because according to the docstring for the function module, this is the only type of function for which we do not have a function body that would otherwise define what assumptions are made. 

Further, the current implementation has the following (possibly confusing) properties:

```python
assert Function('x', real=True) != Function(Symbol('x', real=True))
assert IndexedBase('x', real=True) != IndexedBase(Symbol('x', real=True))
```

For `IndexedBase` the reason is that the symbol/label is part of the equality check, and the inequality is due to `Symbol('x') != Symbol('x', real=True)`. For `Function` it is less clear because only the symbol string is stored in both cases. In this case, it is because the assumptions are not logically expanded before they are stored (via `StdFactKB`). The following is therefore also true after this PR:

```python
x = Symbol('x', real=True)
assert Function('x', **x.assumptions0) == Function(x)
```
Because `Function` did not currently perform assumption expanding prior to this PR, I have not enforced this either (yet). 

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
  * added support for inheriting assumptions from symbols to IndexedBase
* core
  * added support for inheriting assumptions from symbols to Function
<!-- END RELEASE NOTES -->

bsamseth committed:
IndexedBase inherits assumptions from initializing symbol

With reference to discussion PR #16085,
https://github.com/sympy/sympy/pull/16085#discussion_r287332934

bsamseth committed:
Function inherits assumptions if created with Symbol

bsamseth committed:
Redo verification of assumptions

czgdp1807 commented on sympy/core/function.py:
```diff
@@ -863,6 +867,15 @@ class UndefinedFunction(FunctionClass):
+                    raise ValueError(
+                        "clash between assumptions inherited from name and passed assumptions"
+                    )
+                else:
```
I think we can remove this `else`. If  `if` doesn't work then `kwargs[k] = v` will. In other case code will stop execution due to `raise`.

bsamseth replied:
👍

czgdp1807 commented on sympy/tensor/indexed.py:
```diff
@@ -463,7 +471,16 @@ def __new__(cls, label, shape=None, **kw_args):
+                raise ValueError(
+                    "clash between assumptions inherited from label and passed assumptions"
+                )
+            else:
```
+1

czgdp1807 commented on sympy/core/function.py:
```diff
@@ -871,7 +884,7 @@ def __new__(mcl, name, bases=(AppliedUndef,), __dict__=None, **kwargs):
         __dict__.update({'_extra_kwargs': kwargs})
         __dict__['__module__'] = None # For pickling
-        ret = super(UndefinedFunction, mcl).__new__(mcl, name, bases, __dict__)
+        ret = super(UndefinedFunction, mcl).__new__(mcl, str(name), bases, __dict__)
```
Isn't the case that `name` is always a string? As far as I know they are used for printing, right? Is this required for your changes?

bsamseth replied:
It will not be a string when called with `Function(Symbol('f'))`, which is partly what this PR aims to allow. Supplying `Symbol` in place of a string in `Function` constructor (the desired functionality) will not work without some workaround, and this was the simplest way I could come up with.

I guess an argument can be made that because `Function(Symbol('f'))` was not possible before, then this new functionallity should only be applied to `IndexedBase`, leaving `Function` be as it is. Thoughts?

czgdp1807 replied:
I will read the thread here and the related PRs, will try out some experiments.  I will present my views after that. Thanks for the explanation. The work seems interesting.

smichr replied:
The story is that `Symbol('f')(x)` formerly automagically created `Function('f')(x)` but this was frowned upon. The current PR would allow `Function(Symbol(f))` which is about as painless a way to go from Symbols to Functions while retaining the same assumptions.

So what is the use case for this? One would almost always include and argument for the function, e.g. `x -> x(t) ` so to make that change would be either `x.subs(x, Function(x.name, x.assumptions0)(t))` or `x.subs(x, Function(x)(t))`. Are there other use cases? Is there a compelling reason to not allow creation of a a Function from an atom (Symbol) with assumptions?

czgdp1807 replied:
Thanks for providing the summary. I think then this change is fine. I am supportive of this PR. For a simpler way than this I will think and if needed change can be made in future sometime.

bsamseth committed:
Minor simplification to assumption validation

smichr commented on sympy/core/function.py:
```diff
@@ -372,11 +372,15 @@ class Function(Application, Expr):
 
-    Assumptions can be passed to Function.
+    Assumptions can be passed to Function, and if function is initialized with a
+    symbol, the function inherits the associated assumptions:
```
```suggestion
    Symbol, the function inherits the name and assumptions associated with the Symbol:
```

smichr commented on sympy/core/function.py:
```diff
@@ -863,6 +867,14 @@ class UndefinedFunction(FunctionClass):
         __dict__ = __dict__ or {}
         # Allow Function('f', real=True)
+        # and/or Function(Symbol('f', real=True))
+        if hasattr(name, "assumptions0"):
```
```suggestion
        if isinstance(name, Symbol):
```

smichr commented on sympy/core/tests/test_function.py:
```diff
@@ -1100,12 +1100,14 @@ def test_function_assumptions():
 
     assert f(x).is_real is None
     assert f_real(x).is_real is True
+    assert f_real_inherit(x).is_real is True
```
```suggestion
    assert f_real_inherit(x).is_real is True and f_real_inherit.name == 'f'
```

smichr commented on sympy/tensor/indexed.py:
```diff
@@ -410,6 +410,14 @@ class IndexedBase(Expr, NotIterable):
     >>> A != A_real
     True
+
+    Assumptions can also be inherited from the symbol used to initialize the IndexedBase:
```
```suggestion
    Assumptions can also be inherited if a Symbol is used to initialize the IndexedBase:
```

bsamseth committed:
Apply suggestions from code review

Co-Authored-By: Christopher Smith <smichr@gmail.com>

smichr commented on sympy/tensor/indexed.py:
```diff
@@ -463,7 +471,15 @@ def __new__(cls, label, shape=None, **kw_args):
-        assumptions, _ = IndexedBase._filter_assumptions(kw_args)
+
+        assumptions = label.assumptions0
+        passed_assumptions, _ = IndexedBase._filter_assumptions(kw_args)
```
Should this be
```python
if isinstance(label, Symbol):
    assumptions = label.assumptions0
    label = str(label)
else:
    assumptions = {}
```

smichr commented on sympy/tensor/tests/test_indexed.py:
```diff
@@ -222,6 +222,17 @@ def test_IndexedBase_assumptions():
+    assert I_inherit.is_integer
+    assert I_explicit.is_integer
+    assert I_inherit.assumptions0 == I_explicit.assumptions0
+    assert I_inherit != I_explicit  # Because the underlying symbol is different.
```
With the above change, I think these should be the same -- and we want them to be so since once the Symbol passes the "event horizon" it is irrelevant that it was a Symbol. ;-)

ShubhamKJha replied:
I agree with this point, also if the label is supposed to only be a `Symbol`, you can use something like `label = Symbol(label.name, assumptions)` , here `assumptions` that has been filtered for the `IndexedBase`.

bsamseth replied:
The latest commit fixes this, at least in one of many possible ways.

bsamseth committed:
Fix failing test in test_function.py

Comparison was made between a sympy version of 'f' and the actual string 'f'.
The name acts correct, as reflected by the fixed test which now passes.

bsamseth committed:
Fix inequality of identical IndexedBase objects

smichr commented on sympy/core/tests/test_function.py:
```diff
@@ -1100,12 +1100,14 @@ def test_function_assumptions():
 
     assert f(x).is_real is None
     assert f_real(x).is_real is True
+    assert f_real_inherit(x).is_real is True and str(f_real_inherit) == 'f'
```
```
>>> f=Function('f')
>>> f.name
'f'
```
`f_real_inherit.name == 'f'` should work, else something is broken?

bsamseth replied:
Your very right, I thought for a second that it wasn't supposed to be a string because I didn't think I'd made a change that would affect the name. But sure enough, there's a line in the constructor that sets the function name equal to what now can be a symbol. Fixed in the latest commit, hopefully, that also takes care of the failing CI tests.

bsamseth committed:
Fix bug with Function name being set to Symbol and not str

smichr committed:
rework passing of assumptions via symbol

smichr commented:
It was more delicate an operation than I imagined:
```diff
diff --git a/sympy/core/function.py b/sympy/core/function.py
index 83de703..29e77a2 100644
--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -865,40 +865,49 @@ class UndefinedFunction(FunctionClass):
     The (meta)class of undefined functions.
     """
     def __new__(mcl, name, bases=(AppliedUndef,), __dict__=None, **kwargs):
-        __dict__ = __dict__ or {}
+        from .symbol import _filter_assumptions
         # Allow Function('f', real=True)
         # and/or Function(Symbol('f', real=True))
+        assumptions, kwargs = _filter_assumptions(kwargs)
         if isinstance(name, Symbol):
-            for k, v in name.assumptions0.items():
-                if k in kwargs and v != kwargs[k]:
-                    raise ValueError(
-                        "clash between assumptions inherited from name and passed assumptions"
-                    )
-                kwargs[k] = v
-        __dict__.update({'is_' + arg: val for arg, val in kwargs.items() if arg in _assume_defined})
+            assumptions = name._merge(assumptions)
+            name = str(name)
+        elif not isinstance(name, string_types):
+            raise TypeError('expecting string or Symbol for name')
+        else:
+            commutative = assumptions.get('commutative', None)
+            assumptions = Symbol(name, **assumptions).assumptions0
+            if commutative is None:
+                assumptions.pop('commutative')
+        __dict__ = __dict__ or {}
+        # put the `is_*` for into __dict__
+        __dict__.update({'is_%s' % k: v for k, v in assumptions.items()})
         # You can add other attributes, although they do have to be hashable
         # (but seriously, if you want to add anything other than assumptions,
         # just subclass Function)
         __dict__.update(kwargs)
+        # add back the sanitized assumptions without the is_ prefix
+        kwargs.update(assumptions)
         # Save these for __eq__
-        __dict__.update({'_extra_kwargs': kwargs})
-        __dict__['__module__'] = None # For pickling
-        ret = super(UndefinedFunction, mcl).__new__(mcl, str(name), bases, __dict__)
-        ret.name = str(name)
-        return ret
+        __dict__.update({'_kwargs': kwargs})
+        # do this for pickling
+        __dict__['__module__'] = None
+        obj = super(UndefinedFunction, mcl).__new__(mcl, name, bases, __dict__)
+        obj.name = name
+        return obj
 
     def __instancecheck__(cls, instance):
         return cls in type(instance).__mro__
 
-    _extra_kwargs = {}
+    _kwargs = {}
 
     def __hash__(self):
-        return hash((self.class_key(), frozenset(self._extra_kwargs.items())))
+        return hash((self.class_key(), frozenset(self._kwargs.items())))
 
     def __eq__(self, other):
         return (isinstance(other, self.__class__) and
             self.class_key() == other.class_key() and
-            self._extra_kwargs == other._extra_kwargs)
+            self.__dict__ == other.__dict__)
 
     def __ne__(self, other):
         return not self == other
diff --git a/sympy/core/symbol.py b/sympy/core/symbol.py
index b669b96..d2e01c5 100644
--- a/sympy/core/symbol.py
+++ b/sympy/core/symbol.py
@@ -1,6 +1,6 @@
 from __future__ import print_function, division
 
-from sympy.core.assumptions import StdFactKB
+from sympy.core.assumptions import StdFactKB, _assume_defined
 from sympy.core.compatibility import (string_types, range, is_sequence,
     ordered)
 from .basic import Basic
@@ -11,7 +11,7 @@
 from .function import FunctionClass
 from sympy.core.logic import fuzzy_bool
 from sympy.logic.boolalg import Boolean
-from sympy.utilities.iterables import cartes
+from sympy.utilities.iterables import cartes, sift
 from sympy.core.containers import Tuple
 
 import string
@@ -19,6 +19,17 @@
 import random
 
 
+def _filter_assumptions(kwargs):
+    """Split the given dict into assumptions and non-assumptions.
+    Keys are taken as assumptions if they correspond to an
+    entry in ``_assume_defined``.
+    """
+    assumptions, nonassumptions = map(dict, sift(kwargs.items(),
+        lambda i: i[0] in _assume_defined,
+        binary=True))
+    Symbol._sanitize(assumptions)
+    return assumptions, nonassumptions
+
 def _symbol(s, matching_symbol=None, **assumptions):
     """Return s if s is a Symbol, else if s is a string, return either
     the matching_symbol if the names are the same or else a new symbol
@@ -199,6 +210,17 @@ def _sanitize(assumptions, obj=None):
                 continue
             assumptions[key] = bool(v)
 
+    def _merge(self, assumptions):
+        base = self.assumptions0
+        for k in set(assumptions) & set(base):
+            if assumptions[k] != base[k]:
+                raise ValueError(filldedent('''
+                    non-matching assumptions for %s: existing value
+                    is %s and new value is %s''' % (
+                    k, base[k], assumptions[k])))
+        base.update(assumptions)
+        return base
+
     def __new__(cls, name, **assumptions):
         """Symbols are identified by name and assumptions::
 
diff --git a/sympy/core/tests/test_function.py b/sympy/core/tests/test_function.py
index 6e9ffd6..6c8d4f5 100644
--- a/sympy/core/tests/test_function.py
+++ b/sympy/core/tests/test_function.py
@@ -1100,8 +1100,10 @@ def test_function_assumptions():
     x = Symbol('x')
     f = Function('f')
     f_real = Function('f', real=True)
+    f_real1 = Function('f', real=1)
     f_real_inherit = Function(Symbol('f', real=True))
 
+    assert f_real == f_real1  # assumptions are sanitized
     assert f != f_real
     assert f(x) != f_real(x)
 
@@ -1110,7 +1112,8 @@ def test_function_assumptions():
     assert f_real_inherit(x).is_real is True and f_real_inherit.name == 'f'
 
     # Can also do it this way, but it won't be equal to f_real because of the
-    # way UndefinedFunction.__new__ works.
+    # way UndefinedFunction.__new__ works. Any non-recognized assumptions
+    # are just added literally as something which is used in the hash
     f_real2 = Function('f', is_real=True)
     assert f_real2(x).is_real is True
 
diff --git a/sympy/tensor/indexed.py b/sympy/tensor/indexed.py
index e87dbe5..c1c1c78 100644
--- a/sympy/tensor/indexed.py
+++ b/sympy/tensor/indexed.py
@@ -106,8 +106,9 @@
 
 from __future__ import print_function, division
 
-from sympy.core.assumptions import StdFactKB, _assume_defined
-from sympy.core import Expr, Tuple, Symbol, sympify, S
+from sympy.core.assumptions import StdFactKB
+from sympy.core import Expr, Tuple, sympify, S
+from sympy.core.symbol import _filter_assumptions, Symbol
 from sympy.core.compatibility import (is_sequence, string_types, NotIterable,
                                       Iterable)
 from sympy.core.logic import fuzzy_bool
@@ -424,15 +425,6 @@ class IndexedBase(Expr, NotIterable):
     is_Atom = True
 
     @staticmethod
-    def _filter_assumptions(kw_args):
-        """Split the given dict into two parts: assumptions and not assumptions.
-           Keys are taken as assumptions if they correspond to an entry in ``_assume_defined``."""
-        assumptions = {k: v for k, v in kw_args.items() if k in _assume_defined}
-        Symbol._sanitize(assumptions)
-        # return assumptions, not assumptions
-        return assumptions, {k: v for k, v in kw_args.items() if k not in assumptions}
-
-    @staticmethod
     def _set_assumptions(obj, assumptions):
         """Set assumptions on obj, making sure to apply consistent values."""
         tmp_asm_copy = assumptions.copy()
@@ -444,10 +436,12 @@ def _set_assumptions(obj, assumptions):
     def __new__(cls, label, shape=None, **kw_args):
         from sympy import MatrixBase, NDimArray
 
+        assumptions, kw_args = _filter_assumptions(kw_args)
         if isinstance(label, string_types):
             label = Symbol(label)
         elif isinstance(label, Symbol):
-            pass
+            assumptions = label._merge(assumptions)
+            label = Symbol(str(label))
         elif isinstance(label, (MatrixBase, NDimArray)):
             return label
         elif isinstance(label, Iterable):
@@ -463,14 +457,6 @@ def __new__(cls, label, shape=None, **kw_args):
         offset = kw_args.pop('offset', S.Zero)
         strides = kw_args.pop('strides', None)
 
-        # If label is a symbol, ensure the stored label is a plain (no assumptions) symbol,
-        # and propagate any extra assumptions onto the new object.
-        if isinstance(label, Symbol):
-            assumptions = label.assumptions0
-            label = Symbol(str(label))
-        else:
-            assumptions = {}
-
         if shape is not None:
             obj = Expr.__new__(cls, label, shape)
         else:
@@ -480,14 +466,6 @@ def __new__(cls, label, shape=None, **kw_args):
         obj._strides = strides
         obj._name = str(label)
 
-        # Combine assumptions from kw_args and symbol.
-        passed_assumptions, _ = IndexedBase._filter_assumptions(kw_args)
-        for k, v in passed_assumptions.items():
-            if k in assumptions and v != assumptions[k]:
-                raise ValueError(
-                    "clash between assumptions inherited from label and passed assumptions"
-                )
-            assumptions[k] = v
         IndexedBase._set_assumptions(obj, assumptions)
         return obj
 
diff --git a/sympy/utilities/lambdify.py b/sympy/utilities/lambdify.py
index 53039e4..0efa314 100644
--- a/sympy/utilities/lambdify.py
+++ b/sympy/utilities/lambdify.py
@@ -13,6 +13,7 @@
 
 from sympy.core.compatibility import (exec_, is_sequence, iterable,
     NotIterable, string_types, range, builtins, PY3)
+from sympy.utilities.misc import filldedent
 from sympy.utilities.decorator import doctest_depends_on
 
 __doctest_requires__ = {('lambdify',): ['numpy', 'tensorflow']}
@@ -1226,15 +1227,17 @@ def implemented_function(symfunc, implementation):
     # Delayed import to avoid circular imports
     from sympy.core.function import UndefinedFunction
     # if name, create function to hold implementation
-    _extra_kwargs = {}
+    kwargs = {}
     if isinstance(symfunc, UndefinedFunction):
-        _extra_kwargs = symfunc._extra_kwargs
+        kwargs = symfunc._kwargs
         symfunc = symfunc.__name__
     if isinstance(symfunc, string_types):
         # Keyword arguments to UndefinedFunction are added as attributes to
         # the created class.
-        symfunc = UndefinedFunction(symfunc, _imp_=staticmethod(implementation), **_extra_kwargs)
+        symfunc = UndefinedFunction(
+            symfunc, _imp_=staticmethod(implementation), **kwargs)
     elif not isinstance(symfunc, UndefinedFunction):
-        raise ValueError('symfunc should be either a string or'
-                         ' an UndefinedFunction instance.')
+        raise ValueError(filldedent('''
+            symfunc should be either a string or
+            an UndefinedFunction instance.'''))
     return symfunc

```

If you want, you can cherry-pick the commit from my `inherit` branch or apply the diff. It passes symbol, lambdify and indexed tests and doctests.

bsamseth commented:
> It was more delicate an operation than I imagined:

Jeez, now I'm glad I didn't get the time yesterday to try and work it out myself. 

> If you want, you can cherry-pick the commit from my `inherit` branch or apply the diff. It passes symbol, lambdify and indexed tests and doctests.

Cherry-picked the commit, thanks for fixing it!

smichr committed:
Update lambdify.py

IndexedBase('a')[i].subs('a', Dummy('x')) now gives x[i] instead of _x[i] so the routine was changed to use a unique Symbol instead of a unique Dummy. Alternatively, the substitution could be changed so it uses Symbol/Dummy.name instead of str(Symbol/Dummy). Recall that str(Dummy) retains the underscore while str(Symbol) does not. Since we have disambiguate, however, that will highlight differences in Symbols maybe it would just make sense to have Dummy stop preppending symbols with underscores.

smichr committed:
Update lambdify.py

smichr commented:
This has already been open for a few days without comment. I think this makes sense.
